packages/runtime/datastore-definitions/src/channel.ts

```diff
@@ -262,7 +262,7 @@ export interface IChannelServices {
  * the collaborating clients will need to have access to a factory that can produce the `SharedMap` object.
  * @public
  */
-export interface IChannelFactory {
+export interface IChannelFactory<out TChannel extends IFluidLoadable = IFluidLoadable> {
```

Comment:
why constrain to and default as `IFluidLoadable` rather than `IChannel`? Seems like it would give simpler types.

Reply:
I was unable to make the type inference around isDataObject work correctly when I did that (its what I tried first, and I tried it again when making this PR). A lot of code is generic over shared object and data objects, and parameterized by a IFluidLoadable: making shared objects require I channel seems like it should work, but you end up needing an extends clause or another intersection in the LoadableObject union, and that screws up narrowing. Its likley possible to make work, but I tried and failed twice, so I went with this which works fine and is a smaller delta.

Reply:
My experiments with this were before I swapped the order of the LoadableObjectClass union. Let me see if it works better now that that has been swapped.

Reply:
i got it to work with a single cast in RootDataObject.create. its necessary due to limitations of type narrowing on generics:
https://github.com/microsoft/FluidFramework/compare/main...anthony-murphy:prototype-ichannel-in-factory

The end typing is also a bit wonky, as i went with interfaces in the factories, which results in union of class and interface due to the dependency of contianer schema on class. this could be resolve by just using the class as the generic parameter to the factory. it may also be possible to parametrize the factory itself, so it works with either classes or interfaces to keep the ability to more away from classes more easily supported.

Reply:
I tried integrating that change into my [AzureFactory branch](https://github.com/microsoft/FluidFramework/pull/19717), and it broke tons of tests using container schema.

The issue is that if:
```typescript
export type LoadableObjectClass<T extends IFluidLoadable = IFluidLoadable> =
	| (T extends IChannel ? SharedObjectClass<T> : never)
	| DataObjectClass<T>;
```

Then `LoadableObjectClass<IFluidLoadable>` simplifies to DataObjectClass<IFluidLoadable> which means that SharedObjectClass no longer can be used in ContainerSchema.

In short, that version of LoadableObjectClass fails to be covariant with respect to T, which breaks a lot of things including shared objects in container schema.

An interesting feature of the approach I took (`& IChannel` on the return ) is that it now becomes optional for interfaces like ITree to extend IChannel, so we could actually hide (in the container schema based API) the fact that trees expose things like "getAttachSummary", "summarize", "getGCData" if we want to.

Reply:
why make that change at all, can't it remain: 
`export type LoadableObjectClass<T extends IFluidLoadable>`

Reply:
My above comment was in response to Abram's proposal, but @anthony-murphy 's approach:

```typescript
export type LoadableObjectClass<T extends IFluidLoadable> = T extends IChannel
	? SharedObjectClass<T>
	: DataObjectClass<T>;
```

Is one I tried previously, and actually has the same issue where LoadableObjectClass<IFluidLoadable> simplifies to DataObjectClass which means that SharedObjectClass no longer can be used in ContainerSchema when combined with the changes in my [AzureFactory branch](https://github.com/microsoft/FluidFramework/pull/19717).

This change also has the issue that it becomes impossible to properly support a data-object that returns a type which happens to implement IChannel: the old code, my changes and Abram's suggestions all allowed that case. THis change results in my adjusted

```typescript
export function isDataObjectClass<T extends IFluidLoadable>(
	obj: LoadableObjectClass<T>,
): obj is InternalDataObjectClass<T>;
```
failing to compile, and needing to be changed to:
```typescript
export function isDataObjectClass<T extends IFluidLoadable>(
	obj: LoadableObjectClass<T> | DataObjectClass<T>,
): obj is InternalDataObjectClass<T>;
```

If we are willing to keep the use of `any` in ContainerSchema (which results in untyped container schema returning DDSs and shared objects as `any` instead of as `IFluidLoadable` as is remove in my other branch as a follow-up to this work) then I suspect we could make this work, but that seems worse than what I have proposed which avoids the cast, avoids the use of any, and only looks a little funny if you read the implementation of IChannelFactory, but nothing about its unsafe from a type perspective.

The current implementation of InitialObjects does produce `any` with [Anthony](https://github.com/anthony-murphy)'s version, but the one on my branch does work around such issues. Thus it's possible to work around the lack of covariance, but evert user of the container schema type would need to do so, and forgetting to do so will produce `any`  and the workaround is messy and a bit fragile.

I think factory style APIs should maintain covariant and thus be able to avoid `any`. My approach accomplishes this while the other suggestions here do not.

I have moved some of changes (removal of any) from my other branch into this one which benefit from the covariance and added a private doc comment about this design choice.

Reply:
I still think my first proposal builds and solves things for the encapsulated model in isolation. But I didn't realize your other changes actually do introduce typesafety to the azure APIs, at which point I agree the end state is better--seems reasonable to me.